### ISSUE TITLE
How to fix error of "Could not load UVM model"

### ISSUE THREAD

**Issue by qians17** [NVIDIA/nvidia-docker#512]
Hello, 

I have an issue when I install nvidia dockers. When I run the command of "nvidia-docker run --rm nvidia/cuda:8.0-devel nvidia-smi", it will show the error that:
"nvidia-docker | 2017/10/31 11:20:07 Error: Could not load UVM kernel module. Is nvidia-modprobe installed?"

The reason why I set the cuda as cuda:8.0 is because I want to implement DIGITS 6.0 which has a requirement of CUDA 8.0. 
The screenshot of when I run "dpkg -l | grep -i nvidia" is:
ii  bbswitch-dkms                                         0.7-2ubuntu1                                        amd64        Interface for toggling the power on nVidia Optimus video cards
rc  digits                                                4.0.0-1                                             amd64        NVIDIA DIGITS webserver
ii  libcuda1-361                                          361.78-0ubuntu1                                     amd64        NVIDIA CUDA runtime library
rc  libcuda1-384                                          384.90-0ubuntu0.14.04.1                             amd64        NVIDIA CUDA runtime library
ii  libnccl1                                              1.2.1-1+cuda7.5                                     amd64        NVIDIA Communication Collectives Library (NCCL) Runtime
ii  libvisionworks-repo                                   1.6.0.193n                                          amd64        NVIDIA VisionWorks computer vision library.
ii  libvisionworks-sfm-repo                               0.90.1                                              amd64        Package repository for NVIDIA's VisionWorks SFM module.
ii  libvisionworks-tracking-repo                          0.88.0                                              amd64        Package repository for NVIDIA's VisionWorks Tracking module.
ii  nvidia-361                                            361.78-0ubuntu1                                     amd64        NVIDIA binary driver - version 361.78
ii  nvidia-361-dev                                        361.78-0ubuntu1                                     amd64        NVIDIA binary Xorg driver development files
rc  nvidia-384                                            384.90-0ubuntu0.14.04.1                             amd64        NVIDIA binary driver - version 384.90
ii  nvidia-docker                                         1.0.1-1                                             amd64        NVIDIA Docker container tools
ii  nvidia-modprobe                                       361.78-0ubuntu1                                     amd64        Load the NVIDIA kernel driver and create device files
ii  nvidia-opencl-icd-361                                 361.78-0ubuntu1                                     amd64        NVIDIA OpenCL ICD
rc  nvidia-opencl-icd-384                                 384.90-0ubuntu0.14.04.1                             amd64        NVIDIA OpenCL ICD
ii  nvidia-prime                                          0.6.2.1                                             amd64        Tools to enable NVIDIA's Prime
ii  nvidia-settings                                       361.78-0ubuntu1                                     amd64        Tool for configuring the NVIDIA graphics driver

The output when I run "lspci | grep -i nvidia" is:
:00.0 VGA compatible controller: NVIDIA Corporation GM200 [GeForce GTX TITAN X] (rev a1)
05:00.1 Audio device: NVIDIA Corporation Device 0fb0 (rev a1)
06:00.0 VGA compatible controller: NVIDIA Corporation GM200 [GeForce GTX TITAN X] (rev a1)
06:00.1 Audio device: NVIDIA Corporation Device 0fb0 (rev a1)
09:00.0 VGA compatible controller: NVIDIA Corporation GM200 [GeForce GTX TITAN X] (rev a1)
09:00.1 Audio device: NVIDIA Corporation Device 0fb0 (rev a1)
0a:00.0 VGA compatible controller: NVIDIA Corporation GM200 [GeForce GTX TITAN X] (rev a1)
0a:00.1 Audio device: NVIDIA Corporation Device 0fb0 (rev a1)

I think I did install the nvidia-modprobe and I run "nvcc --version" it shows that:
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2016 NVIDIA Corporation
Built on Fri_Jul_15_14:46:45_CDT_2016
Cuda compilation tools, release 8.0, V8.0.33

Could anyone please tell me what's the problem and how to fix that issue?

Thanks,
David

**Comment by flx42** [comment]
Hello, please share the output of:
```
$ sudo nvidia-modprobe -u -c=0
$ echo $?
```

**Comment by qians17** [comment]
The output goes like:
:~$ sudo nvidia-modprobe -u -c=0
modprobe: ERROR: could not insert 'nvidia_361_uvm': Invalid argument

**Comment by flx42** [comment]
Your driver installation seems bogus. Try to `apt-get install nvidia-384`

**Comment by qians17** [comment]
After running that command, the error still remains.

**Comment by flx42** [comment]
Make sure to `apt-get purge` any driver package that has version `361`. Then reboot and try again.

**Comment by lidong04** [comment]
Hi @flx42 , I tried the approach above and #319 , but I still got:
`lidong@SCTCS02:~$ nvidia-docker run -it 83169`
`nvidia-docker | 2017/11/01 21:59:49 Error: Could not load UVM kernel module. Is nvidia-modprobe installed?`

related information is like:
lidong@SCTCS02:~$ docker info
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 43
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-98-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 40
Total Memory: 125.8GiB
Name: SCTCS02
ID: XPMR:WMWJ:GN7R:GUXZ:WX5L:WGMT:GSKU:W777:RFTN:7PZC:B4I6:BAO7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

lidong@SCTCS02:~$ docker version
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:18 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
lidong@SCTCS02:~$ sudo dpkg -l | grep -i nvidia
ii  bbswitch-dkms                              0.8-3ubuntu1                                                amd64        Interface for toggling the power on NVIDIA Optimus video cards
ii  libcuda1-384                               384.90-0ubuntu0.16.04.1                                     amd64        NVIDIA CUDA runtime library
ii  nvidia-384                                 384.90-0ubuntu0.16.04.1                                     amd64        NVIDIA binary driver - version 384.90
ii  nvidia-docker                              1.0.1-1                                                     amd64        NVIDIA Docker container tools
ii  nvidia-opencl-icd-384                      384.90-0ubuntu0.16.04.1                                     amd64        NVIDIA OpenCL ICD
ii  nvidia-prime                               0.8.2                                                       amd64        Tools to enable NVIDIA's Prime
ii  nvidia-settings                            384.90-0ubuntu0~gpu16.04.1                                  amd64        Tool for configuring the NVIDIA graphics driver
lidong@SCTCS02:~$ sudo nvidia-modprobe -u -c=0
sudo: nvidia-modprobe: command not found
lidong@SCTCS02:~$ echo $?
1

**Comment by apolo74** [comment]
Hi guys,
i'm getting the same error but when trying to (re)start one of my containers:
`:~$ sudo nvidia-docker start digits2`
`Error response from daemon: linux runtime spec devices: error gathering device information while adding custom device "/dev/nvidia-uvm-tools": no such file or directory`
`Error: failed to start containers: digits2`

I also tried:

> :~$ sudo nvidia-modprobe -u -c=0

but nothing happened after that, there was no output . Then I tried the `sudo nvidia-docker start digits2` again but I got the same error as before.

I installed nvidia-docker around 3 weeks ago and pulled an *nvidia/digits* image, then I created a container called **digits** and at that moment in time I remember that my nvidia driver was 375.66. At the beginning of this week I realized that Digits was not using the GPU and after many hours of headache I discover that Ubuntu updated the driver to 384.90 and that the container named **digits** was using the *volume* named *nvidia_driver_375.66*. After many more hours of trying to change the container to use the new driver without success, I gave up and decided to lose my information in **digits**. Today I created a new container called **digits2** and everything was working well, Digits recognized the GPU and I could retrain my nets using my GPU... I finished working and decided to *stop* the container, when I tried to restart **digits2** I got the UVM error.

The interesting part is that **digits** (re)start without any problems... it's only **digits2** the one that gives me the UVM error. Any ideas on how to solve this problem besides removing and creating a container every time I want to use the GPU... or maybe a way of making the original **digits** to use the new driver?
Thanks and regards,
Boris

**Comment by lidong04** [comment]
just tried nvidia-docker2 and it worked smoothly :)
@qians17 and @apolo74, you might want to give it a try.

**Comment by apolo74** [comment]
Hi again everybody and thanks for the tip @lidong04 ,
i'm not sure if this is how it should be fixed or not but I hope my workaround is useful for others:
The output of: `sudo nvidia-docker inspect <your_container_name>` gives a JSON list with, among other things, the devices this container should look for... and in my case I saw it like this:
```json
"Devices": [
                {
                    "PathOnHost": "/dev/nvidiactl",
                    "PathInContainer": "/dev/nvidiactl",
                    "CgroupPermissions": "rwm"
                },
                {
                    "PathOnHost": "/dev/nvidia-uvm",
                    "PathInContainer": "/dev/nvidia-uvm",
                    "CgroupPermissions": "rwm"
                },
                {
                    "PathOnHost": "/dev/nvidia-uvm-tools",
                    "PathInContainer": "/dev/nvidia-uvm-tools",
                    "CgroupPermissions": "rwm"
                },
                {
                    "PathOnHost": "/dev/nvidia0",
                    "PathInContainer": "/dev/nvidia0",
                    "CgroupPermissions": "rwm"
                }
            ],
```
At the beginning of the same output you find the path to the configuration of hosts... something like this:
```json
...
        "ResolvConfPath": "/var/lib/docker/containers/<a large folder name here>/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/<a large folder name here>/hostname",
        "HostsPath": "/var/lib/docker/containers/<a large folder name here>/hosts",
        "LogPath": "/var/lib/docker/containe...
```
So I went to that address and found a file named **hostconfig.json** (just in case made a backup copy) and then erased the block calling for the nvidia-uvm-tools device. Then I stopped and restarted *docker* with `sudo systemctl` and finally tried one more time to restart my container with `sudo nvidia-docker start digits2`
It worked and there was no problem to stop and restart it at any time :smile: 
I hope it works for others and I hope to hear from the nvidia-docker crew if this workaround could be valid... in other words, why do I need nvidia-uvm-tools?
I'm guessing that when my Ubuntu updated the nvidia driver to **384.90** the nvidia-docker created containers including this new device... which was not happening for the container created with the previous driver **375.66**.
Cheers,
Boris

**Comment by 3XX0** [comment]
Closing as it most likely a driver installation issue and/or fixed in 2.0/master.
